src/KOKKOS/mliap_so3_kokkos.cpp

```diff
@@ -0,0 +1,1196 @@
+
+       double sa = sinh(rb) / rb;
+       double sb = (cosh(rb) - sa) / rb;
+
+       m_sbes_array[gindex + i1mindex + 0] = sa;
+       m_sbes_array[gindex + i1mindex + 1] = sb;
+
+       int j;
+       for (j = 2; j < t_lmax + 1; j++)
+         m_sbes_array[gindex + i1mindex + j] = m_sbes_array[gindex + i1mindex + j - 2] -
+             (2 * j - 1) / rb * m_sbes_array[gindex + i1mindex + j - 1];
+
+       double exts = m_sbes_array[gindex + i1mindex + j - 2] -
+           (2 * j - 1) / rb * m_sbes_array[gindex + i1mindex + j - 1];
+
+       m_sbes_darray[gindex + i1mindex + 0] = sb;
+
+       for (j = 1; j < t_lmax; j++)
+         m_sbes_darray[gindex + i1mindex + j] = xi *
+             (j * m_sbes_array[gindex + i1mindex + j - 1] +
+              (j + 1) * m_sbes_array[gindex + i1mindex + j + 1]) /
+             (2 * j + 1);
+
+       m_sbes_darray[gindex + i1mindex + j] =
+           xi * (j * m_sbes_array[gindex + i1mindex + j - 1] + (j + 1) * exts) / (2 * j + 1);
+       m_sbes_darray[gindex + i1mindex + 0] = xi * sb;
+     }
+   }
+}
+
+/* ---------------------------------------------------------------------- */
+
+template <class DeviceType>
+KOKKOS_INLINE_FUNCTION
+void MLIAP_SO3Kokkos<DeviceType>::operator() (const MLIAPSO3GetRipArrayTag&, int ii) const{
+   int ipair = t_ij(ii);
+   for (int neighbor = 0; neighbor < t_numneighs[ii]; neighbor++) {
+
+     double x = t_rij(ipair, 0);
+     double y = t_rij(ipair, 1);
+     double z = t_rij(ipair, 2);
+     ipair++;
+
+     double ri = sqrt(x * x + y * y + z * z);
+     if (ri < SMALL) continue;
+
+     double expfac = 4 * MY_PI * exp(-t_alpha * ri * ri);
+     for (int n = 1; n < t_nmax + 1; n++)
+       for (int l = 0; l < t_lmax + 1; l++) {
+         double integral = 0.0, integrald = 0.0;
+         for (int i = 0; i < m_Nmax; i++) {
+           integral += m_g_array[(n - 1) * m_Nmax + i] *
+               m_sbes_array[(ipair - 1) * m_Nmax * (m_lmax + 1) + (bigint) i * (m_lmax + 1) + l];
+           integrald += m_g_array[(n - 1) * m_Nmax + i] *
+               m_sbes_darray[(ipair - 1) * m_Nmax * (m_lmax + 1) + (bigint) i * (m_lmax + 1) + l];
+         }
+
+         m_rip_array[(ipair - 1) * m_nmax * (m_lmax + 1) + (bigint) (n - 1) * (m_lmax + 1) + l] =
+             integral * expfac;
+         m_rip_darray[(ipair - 1) * m_nmax * (m_lmax + 1) + (bigint) (n - 1) * (m_lmax + 1) + l] =
+             integrald * expfac;
+       }
+   }
+}
+
+/* ---------------------------------------------------------------------- */
+
+template <class DeviceType>
+void MLIAP_SO3Kokkos<DeviceType>::spectrum(int nlocal, DAT::tdual_int_1d numneighs, DAT::tdual_int_1d jelems, DAT::tdual_float_1d wjelem,
+                               DAT::tdual_float_2d rij,  DAT::tdual_int_1d k_ij,
+                               int nmax, int lmax, double rcut, double alpha, int totaln, int ncoefs)
+{
+  init_arrays(nlocal, ncoefs);
+
+  bigint totali;
+
+  totali = totaln * m_Nmax * (m_lmax + 1);
```

Comment:
Need to add a check to only realloc if the size has been exceeded. Please also disable initialization if zeroing is not needed.